operator-framework-core/src/test/java/io/javaoperatorsdk/operator/processing/expectation/ExpectationManagerTest.java

```diff
@@ -0,0 +1,208 @@
+  private ConfigMap configMap;
+  private Context<ConfigMap> context;
+
+  @BeforeEach
+  void setUp() {
+    expectationManager = new ExpectationManager<>();
+    configMap = new ConfigMap();
+    configMap.setMetadata(
+        new ObjectMetaBuilder().withName("test-configmap").withNamespace("test-namespace").build());
+    context = mock(Context.class);
+  }
+
+  @Test
+  void setExpectationShouldStoreExpectation() {
+    Expectation<ConfigMap> expectation = mock(Expectation.class);
+    Duration timeout = Duration.ofMinutes(5);
+
+    expectationManager.setExpectation(configMap, timeout, expectation);
+
+    assertThat(expectationManager.isExpectationPresent(configMap)).isTrue();
+    assertThat(expectationManager.getExpectation(configMap)).contains(expectation);
+  }
+
+  @Test
+  void checkExpectationShouldReturnEmptyWhenNoExpectation() {
+    ExpectationResult<ConfigMap> result = expectationManager.checkExpectation(configMap, context);
+
+    assertThat(result.isExpectationPresent()).isFalse();
+  }
+
+  @Test
+  void checkExpectationShouldReturnFulfilledWhenExpectationMet() {
+    Expectation<ConfigMap> expectation = mock(Expectation.class);
+    when(expectation.isFulfilled(configMap, context)).thenReturn(true);
+
+    expectationManager.setExpectation(configMap, Duration.ofMinutes(5), expectation);
+    ExpectationResult<ConfigMap> result = expectationManager.checkExpectation(configMap, context);
+
+    assertThat(result.isExpectationPresent()).isTrue();
+    assertThat(result.isFulfilled()).isTrue();
+    assertThat(result.expectation()).isEqualTo(expectation);
+    assertThat(expectationManager.isExpectationPresent(configMap)).isFalse();
+  }
+
+  @Test
+  void checkExpectationShouldReturnNotFulfilledWhenExpectationNotMet() {
+    Expectation<ConfigMap> expectation = mock(Expectation.class);
+    when(expectation.isFulfilled(configMap, context)).thenReturn(false);
+
+    expectationManager.setExpectation(configMap, Duration.ofMinutes(5), expectation);
+    ExpectationResult<ConfigMap> result = expectationManager.checkExpectation(configMap, context);
+
+    assertThat(result.isExpectationPresent()).isTrue();
+    assertThat(result.isFulfilled()).isFalse();
+    assertThat(result.expectation()).isEqualTo(expectation);
+    assertThat(expectationManager.isExpectationPresent(configMap)).isTrue();
+  }
+
+  @Test
+  void checkExpectationShouldReturnTimedOutWhenExpectationExpired() throws InterruptedException {
+    Expectation<ConfigMap> expectation = mock(Expectation.class);
+    when(expectation.isFulfilled(configMap, context)).thenReturn(false);
+
+    expectationManager.setExpectation(configMap, Duration.ofMillis(1), expectation);
+    Thread.sleep(10);
+    ExpectationResult<ConfigMap> result = expectationManager.checkExpectation(configMap, context);
+
+    assertThat(result.isExpectationPresent()).isTrue();
+    assertThat(result.isTimedOut()).isTrue();
+    assertThat(expectationManager.isExpectationPresent(configMap)).isTrue();
```

Comment:
TBH I find this a little strange. Maybe timedout expectation should be handled differently. Because is it present after it timeouts?

Reply:
yes, but the user then needs to remove timedouted expectation manually?

Reply:
yes, that is the idea